### ISSUE TITLE
consult-yank-pop not working

### ISSUE THREAD

**Issue by xircon** [minad/consult#377]
I am getting an error:
```
Wrong type argument: listp, t
```
Using the default config from the front page.  Any ideas?

Its a clash with ivy:
```
Debugger entered--Lisp error: (wrong-type-argument listp t)
  ivy--update-history(t)
  #<subr ivy-read>("Yank from kill-ring: " #f(compiled-function (str pred action) #<bytecode 0x158e5d6a2603e11e>) :predicate nil :require-match t :initial-input nil :preselect nil :def nil :history t :keymap nil :dynamic-collection nil :extra-props (:caller ivy-completing-read) :caller consult-yank-pop)
  ivy-posframe--read(#<subr ivy-read> "Yank from kill-ring: " #f(compiled-function (str pred action) #<bytecode 0x158e5d6a2603e11e>) :predicate nil :require-match t :initial-input nil :preselect nil :def nil :history t :keymap nil :dynamic-collection nil :extra-props (:caller ivy-completing-read) :caller consult-yank-pop)
  apply(ivy-posframe--read #<subr ivy-read> ("Yank from kill-ring: " #f(compiled-function (str pred action) #<bytecode 0x158e5d6a2603e11e>) :predicate nil :require-match t :initial-input nil :preselect nil :def nil :history t :keymap nil :dynamic-collection nil :extra-props (:caller ivy-completing-read) :caller consult-yank-pop))
  ivy-read("Yank from kill-ring: " #f(compiled-function (str pred action) #<bytecode 0x158e5d6a2603e11e>) :predicate nil :require-match t :initial-input nil :preselect nil :def nil :history t :keymap nil :dynamic-collection nil :extra-props (:caller ivy-completing-read) :caller consult-yank-pop)
  ivy-completing-read("Yank from kill-ring: " #f(compiled-function (str pred action) #<bytecode 0x158e5d6a2603e11e>) nil t nil t nil nil)
  completing-read("Yank from kill-ring: " #f(compiled-function (str pred action) #<bytecode 0x158e5d6a2603e11e>) nil t nil t nil nil)
  #f(compiled-function () #<bytecode -0xf7133979b6345b>)()
  consult--with-preview-1(any #f(compiled-function (cand restore) #<bytecode -0x1a4a908dd2c772e2>) #f(compiled-function (input cand) #<bytecode -0x190c3c3678b6c3d>) #f(compiled-function (&rest args2) #<bytecode 0x6f5b61dd110bbc7>) #f(compiled-function () #<bytecode -0xf7133979b6345b>))
  consult--read-1((";(setq vc-follow-symlinks t)\n" "#+END_SRC \n \n" "#+BEGIN_SRC emacs-lisp \n" "#+BEGIN_SRC emacs-lisp\n(message \"Section 11\")\n" "#+END_SRC \n\n#+BEGIN_SRC emacs-lisp \n" "(ivy-mode)\n(setq ivy-use-virtual-buffers t)\n(setq ..." "     (message \"Section 0\")\n\n" ";;(use-package 'init-keybindings)\n" "(message \"Section 3\")\n\n" "#+END_SRC \n \n\n" "#+END_SRC \n\n\n\n" "    ;; *******************************************..." "#+BEGIN_SRC emacs-lisp " "steve" "  (defun my-yank-pop ()\n      (interactive \"p\")\n  ..." "(defun my-yank-pop ()\n   (interactive \"p\")\n   (con..." "(helm-posframe-enable)\n\n" "Wrong type argument: listp, t\n[2021-07-31T22:19:01..." "+BEGIN_SRC emacs-lisp\n(use-package yankpad\n  :ensu..." "Wrote /home/xircon/dotfiles/emacs.d/recentf [2 tim..." "  (set-face-attribute 'ivy-posframe nil :foregroun..." "(straight-use-package 'general)" "yank-menu-length" "#+TITLE: Notes\n#+AUTHOR: Steve\n#+OPTIONS: num:nil\n..." "* Configuration: \n" "(counsel-yank-pop-preselect-last" "(set-cursor-color \"green\")\n" "Intersect" "shell:\n  program: /bin/bash\n  args:\n    - -l\n    -..." "#! /usr/local/bin/csi -script" "wget wttr.in/Paris.png" "enable_passive: true\npassive_key: CTRL\n" "email   \nxirconuk@gmail.com   \n" "  Descriptive lists will be treated as snippets. Y..." "* Category 1\n" "** Snippet title\n\n    Here's a text snippet I want..." "** Snippet title\n\n    Here's a text snippet I want..." "\n    test" "  - test :: Bananas\n" "https://github.com/Kungsgeten/yankpad" "Descriptive lists will be treated as snippets. You..." "* Category 2\n\n  Descriptive lists will be treated ..." "(use-package yankpad\n  :ensure t\n  :defer 10\n  :in..." "(straight-use-package 'xclip)" "yankpad" "(defun find-fd (&optional dir initial)\n  (interact..." "consult-ripgrep-command\11" "(setq prefix-help-command #'embark-prefix-help-com...") :prompt "Yank from kill-ring: " :history t :sort nil :category consult-yank :require-match t :state #f(compiled-function (cand restore) #<bytecode -0x1a4a908dd2c772e2>) :prompt "Select: " :preview-key any :sort t :lookup #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_123>)
  consult--read((";(setq vc-follow-symlinks t)\n" "#+END_SRC \n \n" "#+BEGIN_SRC emacs-lisp \n" "#+BEGIN_SRC emacs-lisp\n(message \"Section 11\")\n" "#+END_SRC \n\n#+BEGIN_SRC emacs-lisp \n" "(ivy-mode)\n(setq ivy-use-virtual-buffers t)\n(setq ..." "     (message \"Section 0\")\n\n" ";;(use-package 'init-keybindings)\n" "(message \"Section 3\")\n\n" "#+END_SRC \n \n\n" "#+END_SRC \n\n\n\n" "    ;; *******************************************..." "#+BEGIN_SRC emacs-lisp " "steve" "  (defun my-yank-pop ()\n      (interactive \"p\")\n  ..." "(defun my-yank-pop ()\n   (interactive \"p\")\n   (con..." "(helm-posframe-enable)\n\n" "Wrong type argument: listp, t\n[2021-07-31T22:19:01..." "+BEGIN_SRC emacs-lisp\n(use-package yankpad\n  :ensu..." "Wrote /home/xircon/dotfiles/emacs.d/recentf [2 tim..." "  (set-face-attribute 'ivy-posframe nil :foregroun..." "(straight-use-package 'general)" "yank-menu-length" "#+TITLE: Notes\n#+AUTHOR: Steve\n#+OPTIONS: num:nil\n..." "* Configuration: \n" "(counsel-yank-pop-preselect-last" "(set-cursor-color \"green\")\n" "Intersect" "shell:\n  program: /bin/bash\n  args:\n    - -l\n    -..." "#! /usr/local/bin/csi -script" "wget wttr.in/Paris.png" "enable_passive: true\npassive_key: CTRL\n" "email   \nxirconuk@gmail.com   \n" "  Descriptive lists will be treated as snippets. Y..." "* Category 1\n" "** Snippet title\n\n    Here's a text snippet I want..." "** Snippet title\n\n    Here's a text snippet I want..." "\n    test" "  - test :: Bananas\n" "https://github.com/Kungsgeten/yankpad" "Descriptive lists will be treated as snippets. You..." "* Category 2\n\n  Descriptive lists will be treated ..." "(use-package yankpad\n  :ensure t\n  :defer 10\n  :in..." "(straight-use-package 'xclip)" "yankpad" "(defun find-fd (&optional dir initial)\n  (interact..." "consult-ripgrep-command\11" "(setq prefix-help-command #'embark-prefix-help-com...") :prompt "Yank from kill-ring: " :history t :sort nil :category consult-yank :require-match t :state #f(compiled-function (cand restore) #<bytecode -0x1a4a908dd2c772e2>))
  consult--read-from-kill-ring()
  byte-code("\301 \10D\207" [current-prefix-arg consult--read-from-kill-ring] 2)
  consult-yank-pop(1)
  funcall-interactively(consult-yank-pop 1)
  command-execute(consult-yank-pop record)
  counsel-M-x-action("consult-yank-pop")
  ivy-call()
  #<subr ivy-read>("M-x " ("consult-yank-pop" "spa" "yank-pop" "consult-keep-lines" "balance" "company-posframe-mode" "flush-lines" "embark-act" "sort-fields" "straight-pull-all" "straight-use-package" "chess" "yankpad-edit" "yankpad-expand" "ivy-posframe-mode" "fancy-dabbrev-mode" "helm-show-kill-ring" "consult-yank-from-kill-ring" "dired" "counsel-rg" "sort-lines" "bookmark-set" "vterm-toggle" "iy-go-to-char" "consult-buffer" "which-key-mode" "counsel-recentf" "marginalia-mode" "save-place-mode" "straight-version" "shell" "vterm" "ranger" "find-fd" "helm-M-x" "customize" "find-file" "free-keys" "helm-find" "helpful-key" "my-openfile" "consult-grep" "consult-line" "myswitchbuff" "write-region" "list-packages" "vertico-first" "yankpad-insert" "good-scroll-mode" "markdown-preview" ...) :predicate counsel--M-x-externs-predicate :require-match t :history counsel-M-x-history :action counsel-M-x-action :keymap (keymap (67108908 . counsel--info-lookup-symbol) (67108910 . counsel-find-symbol)) :initial-input nil :caller counsel-M-x)
  ivy-posframe--read(#<subr ivy-read> "M-x " ("consult-yank-pop" "spa" "yank-pop" "consult-keep-lines" "balance" "company-posframe-mode" "flush-lines" "embark-act" "sort-fields" "straight-pull-all" "straight-use-package" "chess" "yankpad-edit" "yankpad-expand" "ivy-posframe-mode" "fancy-dabbrev-mode" "helm-show-kill-ring" "consult-yank-from-kill-ring" "dired" "counsel-rg" "sort-lines" "bookmark-set" "vterm-toggle" "iy-go-to-char" "consult-buffer" "which-key-mode" "counsel-recentf" "marginalia-mode" "save-place-mode" "straight-version" "shell" "vterm" "ranger" "find-fd" "helm-M-x" "customize" "find-file" "free-keys" "helm-find" "helpful-key" "my-openfile" "consult-grep" "consult-line" "myswitchbuff" "write-region" "list-packages" "vertico-first" "yankpad-insert" "good-scroll-mode" "markdown-preview" ...) :predicate counsel--M-x-externs-predicate :require-match t :history counsel-M-x-history :action counsel-M-x-action :keymap (keymap (67108908 . counsel--info-lookup-symbol) (67108910 . counsel-find-symbol)) :initial-input nil :caller counsel-M-x)
  apply(ivy-posframe--read #<subr ivy-read> ("M-x " ("consult-yank-pop" "spa" "yank-pop" "consult-keep-lines" "balance" "company-posframe-mode" "flush-lines" "embark-act" "sort-fields" "straight-pull-all" "straight-use-package" "chess" "yankpad-edit" "yankpad-expand" "ivy-posframe-mode" "fancy-dabbrev-mode" "helm-show-kill-ring" "consult-yank-from-kill-ring" "dired" "counsel-rg" "sort-lines" "bookmark-set" "vterm-toggle" "iy-go-to-char" "consult-buffer" "which-key-mode" "counsel-recentf" "marginalia-mode" "save-place-mode" "straight-version" "shell" "vterm" "ranger" "find-fd" "helm-M-x" "customize" "find-file" "free-keys" "helm-find" "helpful-key" "my-openfile" "consult-grep" "consult-line" "myswitchbuff" "write-region" "list-packages" "vertico-first" "yankpad-insert" "good-scroll-mode" "markdown-preview" ...) :predicate counsel--M-x-externs-predicate :require-match t :history counsel-M-x-history :action counsel-M-x-action :keymap (keymap (67108908 . counsel--info-lookup-symbol) (67108910 . counsel-find-symbol)) :initial-input nil :caller counsel-M-x))
  ivy-read("M-x " ("consult-yank-pop" "spa" "yank-pop" "consult-keep-lines" "balance" "company-posframe-mode" "flush-lines" "embark-act" "sort-fields" "straight-pull-all" "straight-use-package" "chess" "yankpad-edit" "yankpad-expand" "ivy-posframe-mode" "fancy-dabbrev-mode" "helm-show-kill-ring" "consult-yank-from-kill-ring" "dired" "counsel-rg" "sort-lines" "bookmark-set" "vterm-toggle" "iy-go-to-char" "consult-buffer" "which-key-mode" "counsel-recentf" "marginalia-mode" "save-place-mode" "straight-version" "shell" "vterm" "ranger" "find-fd" "helm-M-x" "customize" "find-file" "free-keys" "helm-find" "helpful-key" "my-openfile" "consult-grep" "consult-line" "myswitchbuff" "write-region" "list-packages" "vertico-first" "yankpad-insert" "good-scroll-mode" "markdown-preview" ...) :predicate counsel--M-x-externs-predicate :require-match t :history counsel-M-x-history :action counsel-M-x-action :keymap (keymap (67108908 . counsel--info-lookup-symbol) (67108910 . counsel-find-symbol)) :initial-input nil :caller counsel-M-x)
  counsel-M-x()
  funcall-interactively(counsel-M-x)
  command-execute(counsel-M-x)
```
Now to work out where!

**Comment by minad** [comment]
Ivy is not supported by Consult as of now, since Ivy comes with its own Counsel/Swiper libraried (which also were an inspiration for Consult). See #212 for more information. 

Consult works in combination with the libraries Vertico (my recommendation), Selectrum, Incomplete and the default completion UI.

The problem here is that Consult passes the value t as argument for history to *disable* the history. This is supported since Emacs 28 and Ivy probably doesn't support it yet.

**Comment by xircon** [comment]
Got it.  Have been stripping it out (ivy & counsel),  slow work,  but nearly done.

**Comment by oantolin** [comment]
> The problem here is that Consult passes the value t as argument for history to completing-read/read-from-minibuffer to _disable_ the history. This is supported since Emacs 28

Does this mean consult requires Emacs 28?

**Comment by minad** [comment]
Some of the history commands set history=t. I think this breaks completion-sorted-completions (only this function, nothing else) on Emacs < 28. This issue has been present right from the start since I had falsely assumed that history=t is allowed, like for read-from-minibuffer. This was fixed in 28 and I had forgotten about this issue again. So far nobody has complained. The default completion UI/icomplete and history functions don't see much usage it seems. could either repair this by special casing history=t in consult--read for old Emacsen and if default completion is used or better by advising the problematic function.

**Comment by oantolin** [comment]
Oh, I remember this now!

**Comment by minad** [comment]
The advice would be pretty simple, just setting minibuffer-history-variable to some temporary dynamic variable around all-sorted-completions in order to fake disabling the history.

**Comment by minad** [comment]
@oantolin I looked into this again.  The history is disabled by the `consult-history` command, since it does not make sense to keep a history of the history selections. Furthermore sorting is disabled since the history command should preserve the history ordering. Only the combination of enabled sorting plus `minibuffer-history-variable=t` triggers an error in `completion-all-sorted-completions`. Luckily this combination is not used in Consult, making this a non-issue. I think I came to the same conclusion the last time I looked at this and therefore I didn't introduce any fixes/backports in Consult.
